GVFS/GVFS.Common/GitHubUpgrader.cs

```diff
@@ -40,70 +40,129 @@ public partial class ProductUpgrader
             };
 
         private Version installedVersion;
+        private Version newestVersion;
         private Release newestRelease;
         private PhysicalFileSystem fileSystem;
         private ITracer tracer;
 
-        public ProductUpgrader(
+        public GitHubUpgrader(
             string currentVersion,
-            ITracer tracer)
+            ITracer tracer,
+            GitHubUpgraderConfig upgraderConfig)
+            : this(currentVersion, tracer)
+        {
+            this.Config = upgraderConfig;
+        }
+
+        public GitHubUpgrader(string currentVersion, ITracer tracer)
         {
             this.installedVersion = new Version(currentVersion);
             this.fileSystem = new PhysicalFileSystem();
-            this.Ring = RingType.Invalid;
             this.tracer = tracer;
 
-            string upgradesDirectoryPath = GetUpgradesDirectoryPath();
+            string upgradesDirectoryPath = ProductUpgraderInfo.GetUpgradesDirectoryPath();
             this.fileSystem.CreateDirectory(upgradesDirectoryPath);
         }
 
-        public enum RingType
+        public GitHubUpgraderConfig Config { get; private set; }
+
+        public static GitHubUpgrader Create(
+            ITracer tracer,
+            out bool isEnabled,
+            out bool isConfigured,
+            out string error)
         {
-            // The values here should be ascending.
-            // Invalid - User has set an incorrect ring
-            // NoConfig - User has Not set any ring yet
-            // None - User has set a valid "None" ring
-            // (Fast should be greater than Slow,
-            //  Slow should be greater than None, None greater than Invalid.)
-            // This is required for the correct implementation of Ring based
-            // upgrade logic.
-            Invalid = 0,
-            NoConfig = None - 1,
-            None = 10,
-            Slow = None + 1,
-            Fast = Slow + 1,
+            GitHubUpgrader upgrader = null;
+            LocalGVFSConfig localConfig = new LocalGVFSConfig();
+            GitHubUpgraderConfig gitHubUpgraderConfig = new GitHubUpgraderConfig(tracer, localConfig);
+
+            if (gitHubUpgraderConfig.TryLoad(out isEnabled, out isConfigured, out error))
+            {
+                upgrader = new GitHubUpgrader(
+                    ProcessHelper.GetCurrentProcessVersion(),
+                    tracer,
+                    gitHubUpgraderConfig);
+            }
+
+            return upgrader;
         }
 
-        public RingType Ring { get; protected set; }
+        public bool TryInitialize(out string errorMessage)
+        {
+            errorMessage = null;
+            return true;
+        }
 
-        public bool TryGetNewerVersion(
-            out Version newVersion,
-            out string errorMessage)
+        public bool TryGetConfigAllowsUpgrade(out bool upgradeAllowed, out string message)

```

Comment:
I still get caught up on the different aspects this method is checking for, and the different return values...

1) Would `configuredForUpgrade` be a more accurate description of what this method is checking?
2) Could this method only be responsible for determining if it is properly configured for upgrade, and let another method be responsible determining whether there is a misconfiguration or not?

Reply:
Thank you for pointing this out. I received comments from @jrbriggs also about this being difficult to follow. Made the following changes.

* Split the code into two. 1. Loading the config and 2. Deciding whether enough config info is available to do an actual upgrade.
* Updated `GitHubUpgrader.Create()` method to fail when config load fails. When this happens, `ProductUpgraderFactory.TryCreateUpgrader()` will return null with config error messaging. 

Now the callers (`UpgradeVerb`, `UpgradeOrchestrator`, `GVFS.Service`) don't need to check for config load errors. They can use `IProductUpgrader.UpgradeAllowed()` to continue with upgrade process.